helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java

```diff
@@ -306,14 +307,15 @@ public void close() {
     delayedRebalanceSchedule(clusterData, activeNodes, resourceMap.keySet());
 
```

Comment:
After reading through the whole class, I think this is the right place to initialize a new instance of the rebalance algorithm based on the current preference mapping. It could achieve your goal: avoid the race condition, prevent different calculation steps uses different objects and my hope: reduce the number of passed parameters, and remove the private field. Do you think so?

Reply:
I might miss something, but even that's the case, we still need to pass it as a parameter to computeBestPossibleAssignment and applyRebalanceOverwrite to avoid race condition, right? So it just reduces one layer. The cost is that we need to re-create a algorithm on every rebalance.

Another option is refer to the private fields multiple times in this method. And this is also safe for now. But considering the future change, it is safer to call this at just one place.
It is a good practice to reduce refer to a mutable field in the method, you know.